File: app/src/main/java/com/storehop/app/ui/util/UndoEventBus.kt
package com.storehop.app.ui.util

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Process-scoped, single-delivery channel for undo prompts that need to
 * cross screens. Used today for the "Item deleted -- UNDO" snackbar:
 * `ItemFormViewModel` emits after a successful softDelete, and the items
 * list screen (which the form pops back to) collects + shows the snackbar.
 *
 * Why a Channel rather than SharedFlow: each undo prompt should be shown
 * exactly once. Channel.BUFFERED + receiveAsFlow gives us at-most-once
 * delivery -- a second collector wouldn't see events the first already
 * consumed. Within-screen undo (mark purchased, delete store) doesn't go
 * through here; those screens own both the action and the snackbar so
 * local state suffices.
 */
@Singleton
class UndoEventBus @Inject constructor() {

    private val channel = Channel<UndoEvent>(capacity = Channel.BUFFERED)

    val events: Flow<UndoEvent> = channel.receiveAsFlow()

    suspend fun emit(event: UndoEvent) {
        channel.send(event)
    }
}

sealed class UndoEvent {
    /** Item soft-deleted from the form. The list screen offers UNDO. */
    data class ItemDeleted(
        val itemId: String,
        val itemName: String,
    ) : UndoEvent()
}

File: ios/Storehop/UI/Util/UndoEventBus.swift
import Foundation

/// Cross-screen undo events. Today's only producer is `ItemFormViewModel`
/// after a successful `softDelete`; the consumer is `ItemsListViewModel`
/// which the form pops back to and which shows an UNDO snackbar.
///
/// Each event is delivered at most once per subscriber. Within-screen undo
/// (purchase + undo on Shop-at-Store, store delete + undo on Store Picker)
/// doesn't go through here — those screens own both ends and use local
/// state.
enum UndoEvent: Sendable, Hashable {
    /// Item was soft-deleted from the form. The list screen surfaces UNDO.
    case itemDeleted(itemId: String, itemName: String)
}

/// Process-scoped event bus. Backed by an actor so producers and
/// consumers serialize without manual locking.
final actor UndoEventBus {
    private var continuations: [UUID: AsyncStream<UndoEvent>.Continuation] = [:]

    func emit(_ event: UndoEvent) {
        for (_, continuation) in continuations {
            continuation.yield(event)
        }
    }

    nonisolated func events() -> AsyncStream<UndoEvent> {
        AsyncStream { continuation in
            let id = UUID()
            Task { await self.register(id: id, continuation: continuation) }
            continuation.onTermination = { @Sendable _ in
                Task { await self.unregister(id: id) }
            }
        }
    }

    private func register(id: UUID, continuation: AsyncStream<UndoEvent>.Continuation) {
        continuations[id] = continuation
    }

    private func unregister(id: UUID) {
        continuations[id] = nil
    }
}
